src/main/java/org/commonjava/util/sidecar/services/ArchiveRetrieveService.java

```diff
@@ -0,0 +1,279 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commonjava.util.sidecar.services;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.http.NoHttpResponseException;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.conn.ConnectionPoolTimeoutException;
+import org.apache.http.impl.client.BasicCookieStore;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.enterprise.context.ApplicationScoped;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.SocketTimeoutException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+@ApplicationScoped
+public class ArchiveRetrieveService
+{
+
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    public final static String BUILD_CONFIG_ID = "build.config.id";
+
+    private final static int CONNECTION_REQUEST_TIMEOUT = 30 * 60 * 1000; // 30m
+
+    private final static int CONNECTION_TIMEOUT = 30 * 60 * 1000; // 30m
+
+    private final static int SOCKET_TIMEOUT = 30 * 60 * 1000; // 30m
+
+    private final static List<String> decompressedBuilds = new ArrayList<>();
+
+    private final String PART_SUFFIX = ".part";
+
+    private final String ARCHIVE_SUFFIX = ".zip";
+
+    private final String DEFAULT_REPO_PATH = "download";
+
+    @ConfigProperty( name = "archive-api" )
+    public Optional<String> archiveApi;
+
+    @ConfigProperty( name = "local-repository" )
+    private Optional<String> localRepository;
+
+    private CloseableHttpClient client;
+
+    public static boolean isDecompressed()
+    {
+        String buildConfigId = System.getProperty( BUILD_CONFIG_ID );
+        return decompressedBuilds.contains( buildConfigId );
+    }
+
+    @PostConstruct
+    public void init()
+    {
+        final PoolingHttpClientConnectionManager ccm = new PoolingHttpClientConnectionManager();
```

Comment:
I'm thinking if we should try to use new jdk httpclient (https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpClient.html) to exclude a new third-party lib in this sidecar for resource consideration. If we want to use third-party one, maybe jhttpc can be considered here for future metrics support.

Reply:
@ligangty , thanks for this advice, I'd like to see this new httpclient, and I find an interesting page: https://www.mocklab.io/blog/which-java-http-client-should-i-use-in-2020/ There are some comparisons on the normal HTTP clients, and I see Java Httpclient doesn't support Transparent content compression, and Apache supports GZip Deflate, I'm not sure if that will be an issue but I'd like to investigate. Java Httpclient is a new birth and we should open the arms to embrace if possible, I'd like to create a story for this.